templates/Java/libraries/okhttp-gson/ApiClient.mustache

```diff
@@ -0,0 +1,1879 @@
+     *
+     * @param clientId client ID
+     * @param clientSecret client secret
+     * @param parameters a {@link java.util.Map} of parameters
+     */
+    public ApiClient(String clientId, String clientSecret, Map<String, String> parameters) {
+        this(null, clientId, clientSecret, parameters);
+    }
+
+    /**
+     * Constructor for ApiClient to support access token retry on 401/403 configured with base path, client ID, secret, and additional parameters
+     *
+     * @param basePath base path
+     * @param clientId client ID
+     * @param clientSecret client secret
+     * @param parameters a {@link java.util.Map} of parameters
+     */
+    public ApiClient(String basePath, String clientId, String clientSecret, Map<String, String> parameters) {
+        init();
+        if (basePath != null) {
+            this.basePath = basePath;
+        }
+
+        String tokenUrl = "{{{tokenUrl}}}";
+        if (!"".equals(tokenUrl) && !URI.create(tokenUrl).isAbsolute()) {
+            URI uri = URI.create(getBasePath());
+            tokenUrl = uri.getScheme() + ":" +
+                (uri.getAuthority() != null ? "//" + uri.getAuthority() : "") +
+                tokenUrl;
+            if (!URI.create(tokenUrl).isAbsolute()) {
+                throw new IllegalArgumentException("OAuth2 token URL must be an absolute URL");
+            }
+        }
+        RetryingOAuth retryingOAuth = new RetryingOAuth(tokenUrl, clientId, OAuthFlow.{{#lambda.uppercase}}{{#lambda.snakecase}}{{flow}}{{/lambda.snakecase}}{{/lambda.uppercase}}, clientSecret, parameters);
+        authentications.put(
+                "{{name}}",
+                retryingOAuth
+        );
+        initHttpClient(Collections.<Interceptor>singletonList(retryingOAuth));
+        // Setup authentications (key: authentication name, value: authentication).{{#authMethods}}{{#isBasic}}{{#isBasicBasic}}
+        authentications.put("{{name}}", new HttpBasicAuth());{{/isBasicBasic}}{{^isBasicBasic}}
+        authentications.put("{{name}}", new HttpBearerAuth("{{scheme}}"));{{/isBasicBasic}}{{/isBasic}}{{#isApiKey}}
+        authentications.put("{{name}}", new ApiKeyAuth({{#isKeyInHeader}}"header"{{/isKeyInHeader}}{{#isKeyInQuery}}"query"{{/isKeyInQuery}}{{#isKeyInCookie}}"cookie"{{/isKeyInCookie}}, "{{keyParamName}}"));{{/isApiKey}}{{/authMethods}}{{#withAWSV4Signature}}
+        authentications.put("AWS4Auth", new AWS4Auth());{{/withAWSV4Signature}}
+
+        // Prevent the authentications from being modified.
+        authentications = Collections.unmodifiableMap(authentications);
+    }
+
+    {{/-first}}
+    {{/oauthMethods}}
+    {{/hasOAuthMethods}}
+    private void initHttpClient() {
+        initHttpClient(Collections.<Interceptor>emptyList());
+    }
+
+    private void initHttpClient(List<Interceptor> interceptors) {
+        OkHttpClient.Builder builder = new OkHttpClient.Builder();
+        builder.connectTimeout(60, TimeUnit.SECONDS);
```

Comment:
This timeout may not be the one you expect: this controls how long the client will wait for an open socket, and it is only in play _before_ the HTTP request is sent.  Once the socket connection is established, this timeout has no effect.  A timeout here would indicate network connectivity issues between the client and our API...are there really cases where client's can't even get an `ACK` from us within a minute?

The meaning of the different timeouts is explained here: https://www.baeldung.com/okhttp-timeouts

Reply:
Yes, that's correct. I saw the different timeout types and their descriptions, and chose the connect because VMWare was getting a socket timeout exception when the default timeout of 10s was in play. Jinesh Haria asked us to increase the connection timeout for other customers just to avoid that as much as possible.

Personally, I think the SDK is flexible enough that they could use the constructor that allows them to pass in their own OkHttpClient, but because we had a customer issue with a timeout Fabric requested this as a change.

Reply:
CXB Ticket: https://equinixjira.atlassian.net/jira/software/c/projects/CXB/issues/CXB-3465?jql=project%20%3D%20%22CXB%22%20and%20assignee%20%3D%20currentUser%28%29%20ORDER%20BY%20created%20DESC

Reply:
A `SocketTimeoutException` isn't necessarily a connection timeout.  What was the exception message?  (It may be in the ticket you linked, but I don't have access to that.)

Reply:
```
Error:
09/29/2023 12:59:15 :: ------ min: 147ms, max: 3584ms, avg: 317ms, p95: 588ms, p99: 1054ms, p99.9: 3124ms ------
09/29/2023 12:59:16 :: Request POST_SERVICEPROFILE_SEARCH [6321] completed in 261ms
09/29/2023 12:59:17 :: Request POST_SERVICEPROFILE_SEARCH [6322] completed in 255ms
09/29/2023 12:59:18 :: Request POST_SERVICEPROFILE_SEARCH [6323] completed in 233ms
09/29/2023 12:59:29 :: Request [6324] failed after 10005ms
09/29/2023 12:59:29 :: SDK Trace ID: e3c04cd1-fe1b-47d9-8f61-81a665878354
com.equinix.openapi.fabric.ApiException: Message: java.net.SocketTimeoutException: timeout
HTTP response code: 0
HTTP response body: null
HTTP response headers: null
        at com.equinix.openapi.fabric.ApiClient.execute(ApiClient.java:1051)
        at com.equinix.openapi.fabric.v4.api.ServiceProfilesApi.searchServiceProfilesWithHttpInfo(ServiceProfilesApi.java:945)
        at com.equinix.openapi.fabric.v4.api.ServiceProfilesApi.searchServiceProfiles(ServiceProfilesApi.java:923)
        at Example.main(example.java:146)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at jdk.compiler/com.sun.tools.javac.launcher.Main.execute(Main.java:404)
        at jdk.compiler/com.sun.tools.javac.launcher.Main.run(Main.java:179)
        at jdk.compiler/com.sun.tools.javac.launcher.Main.main(Main.java:119)
Caused by: java.net.SocketTimeoutException: timeout
        at okio.SocketAsyncTimeout.newTimeoutException(JvmOkio.kt:147)
        at okio.AsyncTimeout.access$newTimeoutException(AsyncTimeout.kt:158)
        at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:337)
        at okio.RealBufferedSource.indexOf(RealBufferedSource.kt:427)
        at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.kt:320)
        at okhttp3.internal.http1.HeadersReader.readLine(HeadersReader.kt:29)
        at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:178)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at com.equinix.openapi.fabric.ApiClient$2.intercept(ApiClient.java:1458)
```

Reply:
From this it seems that the connect timeout was related to the `SocketTimeoutException`. I'm going to run a quick test and publish the results here.

Reply:
Tested with this script after adding a 60s readTimeout as well: 
(I believe this one should be increased as well to get a general connection+read time to be at least 60s. This ultimately ends up being 1m for connection and 1m for read so could be a total of 2m but it does respond to the initial default timeout increase request.)
```
// Import classes:
import com.equinix.openapi.fabric.ApiClient;
import com.equinix.openapi.fabric.ApiException;
import com.equinix.openapi.fabric.Configuration;
import com.equinix.openapi.fabric.auth.*;
import com.equinix.openapi.fabric.v4.model.*;
import com.equinix.openapi.fabric.v4.api.PortsApi;
import java.util.UUID;

public class Example {
  public static void main(String[] args) {
    ApiClient defaultClient = Configuration.getDefaultApiClient();
//    defaultClient.setBasePath("https://api.equinix.com");
    defaultClient.setBasePath("http://203.0.113.1");
    
    // Configure HTTP bearer authorization: BearerAuth
    HttpBearerAuth BearerAuth = (HttpBearerAuth) defaultClient.getAuthentication("BearerAuth");
    BearerAuth.setBearerToken("<omitted>");

    PortsApi apiInstance = new PortsApi(defaultClient);
    try {
      Port result = apiInstance.getPortByUuid(UUID.fromString("d966b9ce-79ed-4ea6-8f8a-785fd69f3dca"));
      System.out.println(result);
    } catch (ApiException e) {
      System.err.println("Exception when calling ConnectionsApi#createConnection");
      System.err.println("Status code: " + e.getCode());
      System.err.println("Reason: " + e.getResponseBody());
      System.err.println("Response headers: " + e.getResponseHeaders());
      e.printStackTrace();
    }
  }
}
```

That increased the timeout beyond a point where the zScaler gateway would actually timeout without returning a 504:
```
thogarty ~/workspace/fabric-java [CXF-78918-60s-default-timeout] $ java -classpath "equinix-openapi-fabric/target/equinix-openapi-fabric-0.4.0.jar:equinix-openapi-fabric/target/lib/*" thogarty_test_example.java                          
Exception when calling ConnectionsApi#createConnection
Status code: 504
Reason: <html><head><title>504 Gateway Timeout</title></head>
<body><h1>Gateway Timeout</h1>
<p>Server error - server 203.0.113.1 is unreachable at this moment.<br><br>Please retry the request or contact your administrator.<br></p>
<!--Zscaler/6.2--></body></html>

Response headers: {connection=[close], content-type=[text/html], server=[Zscaler/6.2]}
com.equinix.openapi.fabric.ApiException: Message: Gateway Timeout
HTTP response code: 504
HTTP response body: <html><head><title>504 Gateway Timeout</title></head>
<body><h1>Gateway Timeout</h1>
<p>Server error - server 203.0.113.1 is unreachable at this moment.<br><br>Please retry the request or contact your administrator.<br></p>
<!--Zscaler/6.2--></body></html>

HTTP response headers: {connection=[close], content-type=[text/html], server=[Zscaler/6.2]}
	at com.equinix.openapi.fabric.ApiClient.handleResponse(ApiClient.java:1131)
	at com.equinix.openapi.fabric.ApiClient.execute(ApiClient.java:1044)
	at com.equinix.openapi.fabric.v4.api.PortsApi.getPortByUuidWithHttpInfo(PortsApi.java:591)
	at com.equinix.openapi.fabric.v4.api.PortsApi.getPortByUuid(PortsApi.java:569)
	at Example.main(thogarty_test_example.java:22)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.execute(Main.java:404)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.run(Main.java:179)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.main(Main.java:119)
```

Reply:
Actually, just having the connect timeout set to 60s will result in a 120s execution time before receiving the timeout exception though:
```
thogarty ~/workspace/fabric-java [CXF-78918-60s-default-timeout] $ java -classpath "equinix-openapi-fabric/target/equinix-openapi-fabric-0.4.0.jar:equinix-openapi-fabric/target/lib/*" thogarty_test_example.java
connect timeout: 60000
read timeout: 10000
write timeout: 10000
Exception when calling ConnectionsApi#createConnection
Status code: 0
Reason: null
Response headers: null
com.equinix.openapi.fabric.ApiException: Message: java.net.SocketTimeoutException: connect timed out
HTTP response code: 0
HTTP response body: null
HTTP response headers: null
	at com.equinix.openapi.fabric.ApiClient.execute(ApiClient.java:1048)
	at com.equinix.openapi.fabric.v4.api.PortsApi.getPortByUuidWithHttpInfo(PortsApi.java:592)
	at com.equinix.openapi.fabric.v4.api.PortsApi.getPortByUuid(PortsApi.java:570)
	at Example.main(thogarty_test_example.java:26)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.execute(Main.java:404)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.run(Main.java:179)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.main(Main.java:119)
Caused by: java.net.SocketTimeoutException: connect timed out
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:128)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:207)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at com.equinix.openapi.fabric.ApiClient.execute(ApiClient.java:1044)
	... 10 more
	Suppressed: java.net.SocketTimeoutException: connect timed out
		... 34 more
Total execution time: 120089ms
```

Reply:
I'm curious as to which timeouts should be modified to just give the customers some default buffer room.

Reply:
That google url will timeout after 2x the connect timeout. Gateway IP 203.0.113.1 will timeout with whatever the readTimeout will be.

Reply:
Based on the customer error report, the timeout error we're trying to automatically address happens during the HTTP request/response cycle (the original stack trace confirms that, since there are lines in the stack trace like `at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:178)`.  HTTP requires an open TCP connection, so a longer connect timeout would not prevent this kind of error; at the time of opening the connection, there is no HTTP logic involved.  (Also note that the error message reported by the customer is `java.net.SocketTimeoutException: timeout`, whereas the error in your tests is `java.net.SocketTimeoutException: connect timed out`.)  We should not change the default connection timeout because that catches things like critical network failures--or the client pointing at the wrong port, as in your tests--and those are things that should fail fast.

Increasing the read timeout would address the issue in the original report, but setting it to 60 seconds may not do quite what we're aiming for, since the read timeout is still a TCP timeout and not an HTTP timeout.  On top of that, hard-coding that timeout as done here means that if the default _does_ end up being a problem, customers have to explicitly configure their client anyway, so we haven't saved them any work.  If we're going to bake default timeouts into the SDK, then we should also bake in a way to override those defaults by providing a constructor that takes the timeouts as arguments.

We may also want to consider a different set of defaults that what has been proposed so far:

- no change to connect timeout 
- set read and write timeouts to 0
- set call timeout to something slightly longer than the time it would take for something on the Equinix side to return a 504 Gateway Timeout (probably 61 seconds would do it)

With the above defaults, the call timeout implements what people generally think of when they talk about timeouts in the context of an HTTP request: it controls how long the client will wait for the whole process of making the request and reading the response.  Setting read and write timeouts to 0 disables those timeouts, which ensures that the call timeout takes precedence.  Leaving the connect timeout unchanged enables customers to more quickly catch issues where they are fundamentally unable to reach our services at all.

Reply:
Great feedback, Charles! I'll make the updates to the default timeouts as you've suggested:
- no change to connect timeout 
- set read and write timeouts to 0
- set call timeout to something slightly longer than the time it would take for something on the Equinix side to return a 504 Gateway Timeout (probably 61 seconds would do it)

I think the additional constructor may not be necessary because we have a constructor that the customer can use with their own Okhttp client where they can specify their timeouts on there.